README.md
# DAIL2
DAIL simplified for simple minded dev. 

DAIL is a mod compilation for the game Grim Dawn.

The mods in DAIL are there by permission, or the author has an open license. The authors of the mods retain
creative ownership of the content. See the Features_and_Versions.txt file in the Dev_Info folder for a complete
listing of the mods and their authors.

DAIL source is private at the request of mod authors.

_Dev_info/changelog.txt

****************************************************************
****************************************************************
[61.1 6/25/2017]
Update - Coronus

Fix - Voidcaller 
	Chaos Lash projectile values
	(Other skills using tpl need fixed)
	
[61.0 6/24/17]
Fix - Stasher 

[60.9C]
Fix - Installer script and packaging protocols

[60.9B 6/9/2017]
Fix - Removed DGA mastery tag file causing overlap

Fix - Missing _skilltree_consumableskills.dbr entries
	item_cratebox
	dga scroll
	
Fix - Fubar gate
	Removed Bounties Unlimited mod

Fix - GQ Rogue missing UI buttons

Fix - Elementalist icon assets

Fix - Female pc anm entry typo

Add - Grimmest mq_wakingtomissery.qst

[60.9 6/6/2017]
Fix - Demon Hunter tag causing Arch Necro name malfunction

[60.9 6/3/2017]
Fix - Missing GQ sound assets

Update - Cornucopia 401A

Update - DGA 5.1 beta3

[60.9 5/29/2017]
Update - DGA 5.1 beta02b

Update - GrimQuest 1.47

[60.9 4/18/2017]
Update - GrimQuest 1.46HFb

Fix - Defense enumeration

Merge - Bounties Unlocked mod by Ceno
	http://www.grimdawn.com/forums/showthread.php?p=502541#post502541

[60.9 4/17/2017]
Update - GrimQuest 1.46HFa

Fix - Cronley
	Removed old summons to prevent endless heroes

[60.9 4/5/2017]
Update - GrimQuest 1.46

[60.8 3/31/2017]
Fix - Campaign 
	Challenge mode scaling
	
Fix - GQ Epics now show affixes

Fix - Eternal
	Moved aethersoul materia to materia tab

Fix - Fangshi missing skill desc

[60.8 3/29/2017]
Fix - Ranger tag

Fix - Omega Quest
	Alkamos credit
	
[60.8 3/26/2017]
Fix - per AlienFromBeyond
	Various epic/leg affixes

Fix - Missing Grimmest quest edits

Fix - Nature wolf pack buff
	Tier 8 > 7

Fix - Missing Elementalist post 50

Audit - Mastery post 50

Fix - MWar
	Remove - Conversion retalliation on Vessel
	Fix - Lit scaling on damage aura

[60.8 3/24/2017]
Fix - Incorrect X mode proxie pools

Add - XB mode

Fix - TK Evocation ANM

[60.8 3/21/2017]
Fix - Missing skill profile on d001_scepter.dbr

Fix - Missing DGA items outside of DGA

Change - +skill materia ilvl and cost 
	increased to 35, 65, 100
	increased to 1mil, 2mil, 3mil iron

[60.8 3/20/2017]
Update - DGA Beta10b

[60.8 3/19/2017]
Update - per AlienFromBeyond
	Affix tables
	Affixes on crafted items

Update - NCFF 1.33

[60.8 3/172017]
Fix - rare crafting materia drops

Update - Corn 0.3.3I

Update - Coronus 1.5.2

Update - Mystic Warrior

[60.8 3/13/2017]
Repop - Git

Update - Cornucopia 3.3.H

Audit - Modes 
	Proxy pools and challengs
	
Remove - 60.7B release to prevent ass monkey 

[60.7 3/4/2017]
Madness - Hell on wheels
	Fix - component malfunction
	Fix - Omega quest in-op
	Fix - missing Spirit skills
	Update - Cornucopia hotfix
	Update - Grim Quest hotfix
	Update - Dawn of Heros
	
[60.7 3/2/2017]
Update - GrimmerGrimmest
	Proxy pools

Update - Grim Quest 1.4.5

[60.7 2/25/2017]
Begin - DGA beta test releases

[60.7 2/19/2017]
Finish - Mwar 

[60.7 2/16/17]
Update - Lost Treasures 0.40B

Update - Omega 0.78

Update - Cornucopia 0.3.3.E
	Add - GD_Reforge
	
Audit - LCR Loottables

[60.7 2/4/2017]
Change - XP Equation in all Modes
	Old : ((((monsterLevel*12.5-(averagePartyLevel*averagePartyLevel/50))+((monsterLevel-averagePartyLevel)*(averagePartyLevel/1.1)))*0.098)*(1+(monsterExperience/100))*.8)+2
	New (Vanilla) : (((monsterLevel12.75-(averagePartyLevelaveragePartyLevel/50))+((monsterLevel-averagePartyLevel)(averagePartyLevel/1.12)))0.098)*(1+(monsterExperience/100))+3

[60.7 1/28/2017]
Update - Warfare
	Battlestandard - Skill bonus reinstated
	
Update - Terror Knight

[60.7 1/5/2017]
Update - MWar
	PFX - Thanks to Draon Con Color
	
[60.7 12/31/2016]
Update - Lost Treasures 0.33

[60.7 12/30/2016]
Audit - Loot Madness
	Hope it works

[60.7 12/29/2016]
Update - DoH 4.1

Add - additional 'Stuffs' npc
	Griffisu!
		DoH - Custom Relics
		NCFF - Custom Relics

Audit - Enchanter
	Fix - missing stoneguardian tag
	Fix - missing Faerie pet textures 
		In /items, part of loot overhaul

[60.7 12/28/2016]
Audit - Voidcaller
	Change - Tentacle scaled
		0.7 > 0.4
	Known issue - tentacle doesn't always attack
	Change - portrait
	Removed - double icons
	
[60.7 12/27/2016]
Remove - Mode S, SD, SX
	Obelisk onDeath stage 2 spawns
	
[60.7 12/24/2016]
Update - D3 Grimer 

Mod - Mystic Warrior
	Added place holder skills and ui icons

[60.7 12/23/2016]
Update - Omega 0.73

Update - Lost Treasures 0.31

[60.7 12/22/2016]
Update - GQ 1.44A

[60.7 12/17/2016]
Audit - Cenobite
	Fix - Perfect Block
		Animation > BladeBarrier
	Change - Defensive stance	
		Weapon reqs > Shield
	All skills and animations good
	
Audit - Necromancer
	Change - Reapers Touch
		Removed weapon reqs
	All skills and animations good

Audit - Fangshi
	All skills and animations good

[60.7 12/15/2016]
Audit - Ranger
	Conversion completed
	Cleaned up icons
	All skills and animations good

[60.7 12/14/2016]
Change - Continued conversion of NCFF classN to classX

[60.7 12/5/2016]
Change - Cenobite
	Cleaned mastery UI icons

Change - NCFF Tag/UI conversion completed

Update - Necromancer (Enumeration 07)
	Pet folder from newest source

[60.7 12/4/2016]
Change - Begin DAIL second weight reduction pass
	Renamed playerclassN -> playerclassX (In preperation for expansion)
		playerclass07 -> playerclassnecro
		playerclass08 -> playerclassranger
		playerclass10 -> playerclassfangshi
		playerclass14 -> playerclassarnecro
		playerclass28 -> playerclassceno

Removed - All loot tables and chest tables
	Will rebuild tables after Corn and 1007 update

Change - Reduced B and X Mode Density
	Set Boss&Quest Proxies
		spawnMin,1,
		spawnMax,1,
		championMin,2,
		championMax,4,

Added - SD (S - Diminished) Mode
	Set Boss&Quest Proxies
		spawnMin,1,
		spawnMax,1,
		championMin,1,
		championMax,2,
		
[60.6 12/1/2016]
Added - Additional Davood merchandise - AlienFromBeyond 

[60.6 11/29/2016]
Update - Zenith 0.5B

Update - Omega 0.71

Update - Lost Treasures 0.25

Fix - GQ & Ranger anm conflicts

[60.6 11/28/2016]
Post 50 mastery investment inquiries - AlienFromBeyond

Modes/Eternal_Unique/source/scripts/game/wavespawn.lua
/*
	
	GRIM DAWN
	
	For more information, visit us at http://www.grimdawn.com
		
*/

gd.wavespawn = {}
local DebugMode = 0 -- 1 add; 2 remove;
local BossWave = 0
local randomWave = 0
local waveNumber = 0
local spawnPointId = 0
local coords = nil
local chestId = 0
local chestCoords = 0
local shrineId = 0
local shrineCoords = 0
local doorId = 0
local doorCoords = 0
local invisibleWallId = 0
local chestPointCoords = 0
local chestPointId = 0
local shrinePointId = 0
local shrinePointCoords = 0
local spawnProxy = 0
local entityId = 0
local isUpdating = 0
local bossAlive = 0
local waveInProcess = 0
local monsterId = {}
local monstersTotal = 0
local waveTime = 0
local Task_Trial = {}
local Quest_CompleteTrials = 0xD8A2C400
Task_Trial[1] = 0xA600A800
Task_Trial[2] = 0xF9494F00
Task_Trial[3] = 0xCC87B400
Task_Trial[4] = 0x43621F00
Task_Trial[5] = 0x4ACA5200
Task_Trial[6] = 0x75260580
Task_Trial[7] = 0x5B91DA80
Task_Trial[8] = 0x4603BB00
Task_Trial[9] = 0x821C1200
Task_Trial[10] = 0x6F480980
Task_Trial[11] = 0x2F629080
Task_Trial[12] = 0xC455B900
Task_Trial[13] = 0x1C0B2000
Task_Trial[14] = 0x550B5E00
Task_Trial[15] = 0xD719C100
Task_Trial[16] = 0x4D575E80
function gd.wavespawn.Init()
		gd.wavespawn.unlockDoor()
		gd.wavespawn.updateQuest()
		gd.wavespawn.ApplyTokens()
end

function gd.wavespawn.ApplyTokens()
local player = Game.GetLocalPlayer()
	
	if (player:HasToken("DISMANTLING_UNLOCKED") == false) then
		player:GiveToken("DISMANTLING_UNLOCKED")
	end
/*	FACTION EDIT 	
	if (player:HasToken("FACTIONS_SET_FIX3") == false) then
		player:GrantQuest(0x8F577200, 0xB6143900)
		player:GetQuestTaskState(0x8F577200, 0xB6143900)
		player:GiveToken("FACTIONS_SET_FIX3")
	end
*/	
	if (player:HasToken("ASCEND_0") == false) then
		player:GiveToken("ASCEND_0")
	end
	
	if (player:HasToken("STARTING_ITEMS") == false) then
		player:GiveToken("STARTING_ITEMS")
		player:GiveItem("records/items/misc/potions/potion_energya01.dbr", 5, false)
		player:GiveItem("records/items/misc/potions/potion_healtha01.dbr", 5, false)
		player:AdjustMoney(1000)
	end
	if (DebugMode == 1) then
		player:GiveToken("WAVE_45")
		player:GiveToken("DEBUG")
		player:GiveLevels(84)
		player:GiveSkillPoints(100)
		player:AdjustMoney(10000000)
	elseif (DebugMode == 2) then
		player:RemoveToken("DEBUG")
		player:RemoveToken("WAVE_45")
	elseif (DebugMode == 3) then
		player:RemoveToken("WAVE_45")
		player:GiveToken("WAVE_44")
		player:GiveToken("DEBUG")
	end
	
	
end

function gd.wavespawn.updateQuest()
local player = Game.GetLocalPlayer()
local i = 1
	while i <= table.getn(Task_Trial) do
		local quest_state = player:GetQuestTaskState(Quest_CompleteTrials, Task_Trial[i])
		if (quest_state != QuestState.Complete) then
			player:GrantQuest(Quest_CompleteTrials, Task_Trial[i])
		end
		quest_state = player:GetQuestTaskState(Quest_CompleteTrials, Task_Trial[i])
		if (quest_state != QuestState.Complete) then break end
		i=i+1
	end
end

function gd.wavespawn.GetWaveProcess()
	return waveInProcess, nil
end

function gd.wavespawn.SetWaveNumber(n)
	waveNumber = n-1
	UI.Notify("tagWaveToSpawn"..tostring(waveNumber+1))
end

function gd.wavespawn.creatureSpawnPointCoords(objectId)
gd.wavespawn.Init()
	spawnPointId = Entity.Get(objectId)
	coords = spawnPointId:GetCoords()
end

function gd.wavespawn.regChestPoint(objectId)
	chestPointId = Entity.Get(objectId)
	chestPointCoords = chestPointId:GetCoords()
end

function gd.wavespawn.regShrinePoint(objectId)
	shrinePointId = Entity.Get(objectId)
	shrinePointCoords = shrinePointId:GetCoords()
end

function gd.wavespawn.ClearTokens()
local player = Game.GetLocalPlayer()

player:RemoveToken("BOSS_WARDEN_KILLED")
player:RemoveToken("BOSS_DARIUS_KILLED")
player:RemoveToken("BOSS_BLOODFEAST_KILLED")
player:RemoveToken("BOSS_RAVNA_KILLED")
player:RemoveToken("BOSS_OBELISK_KILLED")
player:RemoveToken("BOSS_ZEALOT_KILLED")
player:RemoveToken("BOSS_LOGHORREAN_KILLED")
local i = 1
	while (i <= 100) do
		player:RemoveToken("SHRINE_"..tostring(i))
		player:RemoveToken("WAVE_"..tostring(i))
		i=i+1
	end
waveNumber = 0
player:RemoveToken("GOD_ALKAMOS_KILLED")
player:RemoveToken("GOD_DEPRAVITIE_KILLED")
player:RemoveToken("GOD_ZAULKAHN_KILLED")
player:RemoveToken("GOD_URZOG_KILLED")
player:RemoveToken("GOD_SABRINA_KILLED")
player:RemoveToken("GOD_ARIA_KILLED")
player:RemoveToken("GOD_BEANIES_KILLED")
player:RemoveToken("GOD_HARUN_KILLED")
player:RemoveToken("GOD_MOGDROGEN_KILLED")
player:RemoveToken("GOD_RAZNAR_KILLED")

player:RemoveToken("WEAPON_LOW")
player:RemoveToken("WEAPON_MID")
end

function gd.wavespawn.warningWavespawn()
if (waveInProcess == 1) then return end
	waveNumber = waveNumber+1
	waveTime = 0
	gd.wavespawn.fillArea()
	UI.Notify("tagWave"..tostring(waveNumber))
	if (chestId != 0) then gd.wavespawn.removeChest() end
	if (shrineId != 0) then gd.wavespawn.removeShrine() end
	gd.wavespawn.lockDoor()
	if (isUpdating == 0) then gd.wavespawn.regUpdate(entityId) end
end

function gd.wavespawn.fillArea()
--ASCEND
local player = Game.GetLocalPlayer()
	local i = 0
	while i <= 100 do
		if player:HasToken("ASCEND_"..tostring(i)) then
			i=i+1
		else break end			
	end
	
	if i > 2 then i = 2 end -- temp limit
	i=i-1
	if i > 0 then
		spawnProxy = Proxy.Create("records/proxies/ascend_"..tostring(i).."/proxy_wave"..tostring(waveNumber)..".dbr")
		if (spawnProxy == nil) then
			spawnProxy = Proxy.Create("records/proxies/ascend_"..tostring(i).."/proxy_wave"..tostring(waveNumber)..".dbr")
			UI.Notify("tagNoMoreWaves")
			gd.wavespawn.ClearTokens()
		end
		spawnProxy:SetCoords(coords)
		waveInProcess = 1
		gd.wavespawn.ApplyTokens()
		return
	end
	--ASCEND
	
	spawnProxy = Proxy.Create("records/proxies/proxy_wave"..tostring(waveNumber)..".dbr")
	if (spawnProxy == nil) then
		spawnProxy = Proxy.Create("records/proxies/proxy_wave"..tostring(waveNumber)..".dbr")
		UI.Notify("tagNoMoreWaves")
		gd.wavespawn.ClearTokens()
	end
	spawnProxy:SetCoords(coords)
	waveInProcess = 1
	gd.wavespawn.ApplyTokens()
end

function gd.wavespawn.randomWarningWavespawn()
if (waveInProcess == 1) then return end
	waveTime = 0
	gd.wavespawn.randomFillArea()
	UI.Notify("tagWaveRandom")
	if (chestId != 0) then gd.wavespawn.removeChest() end
	if (shrineId != 0) then gd.wavespawn.removeShrine() end
	gd.wavespawn.lockDoor()
	if (isUpdating == 0) then gd.wavespawn.regUpdate(entityId) end
	randomWave = 1
end

function gd.wavespawn.randomFillArea()
math.randomseed(Time.Now())
local i = random(1,7)
		spawnProxy = Proxy.Create("records/proxies/proxy_wave_random"..tostring(i)..".dbr")
		spawnProxy:SetCoords(coords)
		waveInProcess = 1
end

function gd.wavespawn.regUpdate(objectId)
	entityId = objectId
	Script.RegisterForUpdate(entityId, "gd.wavespawn.Proxy_Update", 1000)
	isUpdating = 1
end

function gd.wavespawn.Proxy_Update(objectId)
local player = Game.GetLocalPlayer()
waveTime = waveTime+1
	if (spawnProxy:AllKilled()) && (bossAlive == 0) && spawnProxy != 0 then
		Script.UnregisterForUpdate(objectId, "gd.wavespawn.Proxy_Update")
		spawnProxy:Destroy()
		spawnProxy = 0
		gd.wavespawn.unlockDoor()
		player:GiveToken("WAVE_"..tostring(waveNumber))
		gd.wavespawn.updateQuest()
		UI.Notify("tagClear")
		isUpdating = 0
		waveInProcess = 0
		monstersTotal = 0
		gd.wavespawn.timeReward()
		waveTime = 0
		if (randomWave != 1) then gd.wavespawn.spawnShrine() end
		randomWave = 0
		BossWave = 0
	end
end

function gd.wavespawn.CreateEffect(coords)
	local pfx = Entity.Create("records/fx/ui/riftgatepersonalopen_fxpak.dbr")
	if pfx != nil then
		pfx:NetworkEnable()
		pfx:SetCoords(coords)	
	end	
end

function gd.wavespawn.timeReward()
local player = Game.GetLocalPlayer()
local avgLevel = Game.GetAveragePlayerLevel()
local rank = 0
	if (waveTime <= 300) then
		UI.Notify("tagWaveTime"..tostring(waveTime))
	else
		UI.Notify("tagWaveTimeMany")
	end
	
	if (randomWave != 1) then
		local i = 0 --Checking time record
		while (i <= 300) do
			if (player:HasToken("WAVE_"..tostring(waveNumber).."TIME-"..tostring(i))) then
				UI.Notify("tagWAVE_"..tostring(waveNumber).."_TIME-"..tostring(i))
				break
			end
			i=i+1
		end
		if (waveTime <= 300) then
			player:GiveToken("WAVE_"..tostring(waveNumber).."TIME-"..tostring(waveTime))
		end
	end
	
	if (BossWave == 0) then
		if (waveTime <= 45) then
			rank = 8
		elseif (waveTime > 45 && waveTime <= 70) then
			rank = 7
		elseif (waveTime > 70 && waveTime <= 80) then
			rank = 6
		elseif (waveTime > 80 && waveTime <= 90) then
			rank = 5
		elseif (waveTime > 90 && waveTime <= 100) then
			rank = 4
		elseif (waveTime > 100 && waveTime <= 110) then
			rank = 3
		elseif (waveTime > 110 && waveTime <= 120) then
			rank = 2
		elseif (waveTime > 120) then
			rank = 1
		end
	elseif (BossWave == 1) then
		if (waveTime <= 90) then
			rank = 8
		elseif (waveTime > 90 && waveTime <= 140) then
			rank = 7
		elseif (waveTime > 140 && waveTime <= 160) then
			rank = 6
		elseif (waveTime > 160 && waveTime <= 180) then
			rank = 5
		elseif (waveTime > 180 && waveTime <= 200) then
			rank = 4
		elseif (waveTime > 200 && waveTime <= 220) then
			rank = 3
		elseif (waveTime > 220 && waveTime <= 240) then
			rank = 2
		elseif (waveTime > 240) then
			rank = 1
		end
	end
	
	local multiplier = 0
	if (Game.GetGameDifficulty() == Game.Difficulty.Normal) then
		multiplier = 1
	elseif (Game.GetGameDifficulty() == Game.Difficulty.Epic) then
		multiplier = 3
	elseif (Game.GetGameDifficulty() == Game.Difficulty.Legendary) then
		multiplier = 5
	end
	UI.Notify("tagRank"..tostring(rank))
	local money = ((30*(rank/2)*waveNumber*avgLevel)/6)*multiplier --was 50
	local experience = ((55+(avgLevel*2))*(rank/2)*waveNumber)*multiplier --was 75
	
	player:AdjustMoney(money)
	player:GiveExperience(experience)
end

function gd.wavespawn.heroDeath()
if (waveTime == 0) then return end
if (chestId != 0) then return end
	local chestDBR = "records/items/lootchests/chest_t0.dbr"
	local chest = Chest.Create(chestDBR)
	chest:SetCoords(chestPointCoords)
end

function gd.wavespawn.spawnShrine()
local player = Game.GetLocalPlayer()
	if( math.mod( waveNumber, 2 ) == 0 ) then
		if (player:HasToken("SHRINE_"..tostring(waveNumber)) == false) then
			local shrine = Shrine.Create("records/interactive/void_devotionshrine.dbr")
			shrine:SetCoords(shrinePointCoords)
		end
	end
end

function gd.wavespawn.useShrine()
local player = Game.GetLocalPlayer()
	player:GiveToken("SHRINE_"..tostring(waveNumber))	
	math.randomseed(Time.Now())
	local i = random(1,100)
		if (i >= 75) then
			UI.Notify("tagAttributePointAwardedc")
			player:GiveItem("records/items/misc/essence_power.dbr", 1, true)
			return
		end
		
		if (i >= 50) then
			UI.Notify("tagSkillPointAwardedc")
			player:GiveItem("records/items/misc/essence_skill.dbr", 1, true)
		end
end

function gd.wavespawn.onSpawnChest(objectId)
	chestId = Chest.Get(objectId)
	chestCoords = chestId:GetCoords()
	local pfx = Entity.Create("records/fx/whiteflash01.dbr")
	if pfx != nil then
		pfx:NetworkEnable()
		pfx:SetCoords(chestCoords)
	end
end

function gd.wavespawn.removeChest()
if (chestId == 0) then return end
	local pfx = Entity.Create("records/fx/whiteflash01.dbr")
	if pfx != nil then
		pfx:NetworkEnable()
		pfx:SetCoords(chestCoords)
	end
	chestId:Destroy()
	chestId = 0
end

function gd.wavespawn.onDoorSpawn(objectId)
	doorId = Door.Get(objectId)
	doorCoords = doorId:GetCoords()
end

function gd.wavespawn.unlockDoor()
if (doorId == 0) then return end
	doorId:SetLocked(false)
	doorId:Open()
		if (invisibleWallId != 0) then	
			invisibleWallId:Destroy()
			invisibleWallId = 0	
		end	
end

function gd.wavespawn.lockDoor()
	doorId:SetLocked(true)
	doorId:Close()
	invisibleWallId = Destructible.Create("records/level art/doorsinterior/invisible_wall.dbr")
	invisibleWallId:SetCoords(doorCoords)
	doorId:SetCoords(doorCoords)
end

function gd.wavespawn.onSpawnShrine(objectId)
	shrineId = Shrine.Get(objectId)
	shrineCoords = shrineId:GetCoords()
	local pfx = Entity.Create("records/fx/whiteflash01.dbr")
	if pfx != nil then
		pfx:NetworkEnable()
		pfx:SetCoords(shrineCoords)
	end
end

function gd.wavespawn.removeShrine()
if (shrineId == 0) then return end
	local pfx = Entity.Create("records/fx/whiteflash01.dbr")
	if pfx != nil then
		pfx:NetworkEnable()
		pfx:SetCoords(shrineCoords)
	end
	shrineId:Destroy()
	shrineId = 0
end

function gd.wavespawn.resetWaves()
	Script.UnregisterForUpdate(entityId, "gd.wavespawn.Proxy_Update")
	if (spawnProxy != 0) then spawnProxy:Destroy() end	
	isUpdating = 0
	waveNumber = 0
	bossAlive = 0
	waveInProcess = 0
	waveTime = 0
	randomWave = 0
	gd.wavespawn.unlockDoor()
	if (chestId != 0) then gd.wavespawn.removeChest() end
	if (shrineId != 0) then gd.wavespawn.removeShrine() end
	UI.Notify("tagClearing")
	local i = 1
	while (i <= monstersTotal) do
		local monCoords = monsterId[i]:GetCoords()
		gd.wavespawn.CreateEffect(monCoords)
		monsterId[i]:Destroy()
		monsterId[i]:Kill()
		i = i+1
	end
	monstersTotal = 0
end

function gd.wavespawn.setBossAlive(objectId)
	bossAlive = 1
	BossWave = 1
	gd.wavespawn.regMonster(objectId)
end

function gd.wavespawn.unsetBossAlive(objectId) --Boss Died
bossAlive = 0
local player = Game.GetLocalPlayer()
local boss = Character.Get(objectId)
local dbrPath = boss:GetName()
local bossName = ""
	if (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_warden02.dbr") then bossName = "WARDEN"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_darius.dbr") then bossName = "DARIUS"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_dranghoul.dbr") then bossName = "BLOODFEAST"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_dermapteranqueen02.dbr") then bossName = "RAVNA"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_aetherialobelisk02.dbr") then bossName = "OBELISK"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_zealot.dbr") then bossName = "ZEALOT"
	elseif (dbrPath == "records/creatures/wavespawn/heroes&bosses/boss_loghorrean_03.dbr") then bossName = "LOGHORREAN"
	end
	if (player:HasToken("BOSS_"..bossName.."_KILLED") == false) then
		chestDBR = "records/items/lootchests/chest_"..bossName..".dbr"
		if (player:HasToken("SKILLPOINT_"..bossName) == false) then
			player:GiveItem("records/items/misc/essence_skill.dbr", 1, true)
			UI.Notify("tagSkillPointAwardedc")
			player:GiveItem("records/items/misc/inventorybag.dbr", 1, false)
			player:GiveToken("SKILLPOINT_"..bossName)
		end
		player:GiveToken("BOSS_"..bossName.."_KILLED")
	else chestDBR = "records/items/lootchests/chest_t0.dbr"
	end
	local chest = Chest.Create(chestDBR)
	chest:SetCoords(chestPointCoords)
end

function gd.wavespawn.regMonster(objectId)
	monstersTotal = monstersTotal+1
	monsterId[monstersTotal] = Character.Get(objectId)
end

local function CharacterCreateArgs()
	local averageLevel = Game.GetAveragePlayerLevel()
	local monsterLevel = (averageLevel+15) --+15

	return monsterLevel, nil	
end

function gd.wavespawn.spawnGod(godName)
	local spawn = Character.Create ("records/creatures/gods/"..godName..".dbr", CharacterCreateArgs())
	spawn:SetCoords(coords)
	monstersTotal = monstersTotal+1
	monsterId[monstersTotal] = spawn
	gd.wavespawn.lockDoor()
	isUpdating = 1
	waveInProcess = 1
	randomWave = 1 --block shrine spawn just in case
	UI.Notify("tagGodSpawned")
end

function gd.wavespawn.godDeath(objectId)
local player = Game.GetLocalPlayer()
local god = Character.Get(objectId)
local dbrPath = god:GetName()
local godName = ""
	if (dbrPath == "records/creatures/gods/alkamos.dbr") then godName = "ALKAMOS"
	elseif (dbrPath == "records/creatures/gods/depravitie.dbr") then godName = "DEPRAVITIE"
	elseif (dbrPath == "records/creatures/gods/zaulkahn.dbr") then godName = "ZAULKAHN"
	elseif (dbrPath == "records/creatures/gods/urzog.dbr") then godName = "URZOG"
	elseif (dbrPath == "records/creatures/gods/sabrina.dbr") then godName = "SABRINA"
	elseif (dbrPath == "records/creatures/gods/aria.dbr") then godName = "ARIA"
	elseif (dbrPath == "records/creatures/gods/beanies.dbr") then godName = "BEANIES"
	elseif (dbrPath == "records/creatures/gods/harun_skeleton.dbr") then godName = "HARUN"
	elseif (dbrPath == "records/creatures/gods/mogdrogen.dbr") then godName = "MOGDROGEN"
	elseif (dbrPath == "records/creatures/gods/raznar.dbr") then godName = "RAZNAR"
	end
	player:GiveToken("GOD_"..godName.."_KILLED")	
	isUpdating = 0
	waveInProcess = 0
	randomWave = 0
	UI.Notify("tagGodKilled")
	gd.wavespawn.unlockDoor()
	--player:GiveExperience(115000)
end

function gd.wavespawn.giveGodItem(itemName)
	local player = Game.GetLocalPlayer()
	player:GiveItem("records/items/godly/"..itemName..".dbr", 1, true)
end

function gd.wavespawn.Ascend()
local player = Game.GetLocalPlayer()
local i = 1
	while i <= 1 do
		if player:HasToken("ASCEND_"..tostring(i)) then
			i=i+1
		else break end
	end
	if i > 1 then i = 1 end
	gd.wavespawn.ClearTokens()
	player:GiveToken("ASCEND_"..tostring(i))
	UI.Notify("tagAscend_"..tostring(i))
end

function gd.wavespawn.Descend()
local player = Game.GetLocalPlayer()
local i = 1
	while i <= 100 do
		player:RemoveToken("ASCEND_"..tostring(i))
			i=i+1
	end
	UI.Notify("tagDescend")
end